File: app/src/main/java/com/siju/acexplorer/settings/SettingsPreferenceFragment.java
/*
 * Copyright (C) 2017 Ace Explorer owned by Siju Sakaria
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.siju.acexplorer.settings;

import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.preference.CheckBoxPreference;
import androidx.preference.ListPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.PreferenceManager;

import com.siju.acexplorer.AceApplication;
import com.siju.acexplorer.R;
import com.siju.acexplorer.analytics.Analytics;
import com.siju.acexplorer.logging.Logger;
import com.siju.acexplorer.main.model.FileConstants;
import com.siju.acexplorer.main.model.helper.FileUtils;
import com.siju.acexplorer.theme.ThemeUtils;
import com.siju.acexplorer.utils.LocaleHelper;

import static com.siju.acexplorer.theme.ThemeUtils.CURRENT_THEME;
import static com.siju.acexplorer.theme.ThemeUtils.PREFS_THEME;


public class SettingsPreferenceFragment extends PreferenceFragmentCompat  {

    private final String TAG = this.getClass().getSimpleName();
    public static final String PREFS_LANGUAGE = "prefLanguage";
    public static final String PREFS_ANALYTICS = "prefsAnalytics";

    private SharedPreferences preferences;
    private Preference updatePreference;
    private String currentLanguage;
    private int theme;

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        setPreferencesFromResource(R.xml.pref_settings, rootKey);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        preferences = PreferenceManager.getDefaultSharedPreferences(getActivity());

        ListPreference langPreference = findPreference(PREFS_LANGUAGE);
        ListPreference themePreference = findPreference(PREFS_THEME);
        theme = ThemeUtils.getTheme(getActivity());
        String PREFS_UPDATE = "prefsUpdate";
        updatePreference = findPreference(PREFS_UPDATE);


        CheckBoxPreference hiddenPreference = findPreference("prefHidden");
        hiddenPreference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                return true;
            }
        });


        CheckBoxPreference analyticsPreference = findPreference(PREFS_ANALYTICS);
        analyticsPreference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                Analytics.getLogger().sendAnalytics((Boolean) newValue);
                return true;
            }
        });


        Preference resetPreference = findPreference(FileConstants.PREFS_RESET);
        resetPreference.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                Intent data = new Intent();
                data.putExtra(FileConstants.PREFS_RESET, true);
                Toast.makeText(AceApplication.Companion.getAppContext(), getString(R.string.msg_fav_reset), Toast
                        .LENGTH_LONG).show();
                getActivity().setResult(AppCompatActivity.RESULT_OK, data);
                return false;
            }
        });

        String value = LocaleHelper.getLanguage(getActivity());
        langPreference.setValue(value);
        currentLanguage = value;
        Logger.log("Settings", "lang=" + currentLanguage);


        // Bind the summaries of EditText/List/Dialog/Ringtone preferences
        // to their values. When their values change, their summaries are
        // updated to reflect the new value, per the Android Design
        // guidelines.
        bindPreferenceSummaryToValue(langPreference);
        bindPreferenceSummaryToValue(themePreference);
        initializeListeners();
    }

    private void initializeListeners() {
        updatePreference
                .setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {

                    @Override
                    public boolean onPreferenceClick(Preference preference) {
                        Intent intent = new Intent(Intent.ACTION_VIEW);
                        // Try Google play
                        intent.setData(Uri
                                .parse("market://details?id=" + getActivity().getPackageName()));

                        if (FileUtils.isPackageIntentUnavailable(getActivity(), intent)) {
                            // Market (Google play) app seems not installed,
                            // let's try to open a webbrowser
                            intent.setData(Uri.parse("https://play.google.com/store/apps/details?id=" +
                                    getActivity().getPackageName()));
                            if (FileUtils.isPackageIntentUnavailable(getActivity(), intent)) {
                                Toast.makeText(getActivity(),
                                        getString(R.string.msg_error_not_supported),
                                        Toast.LENGTH_SHORT).show();
                            } else {
                                startActivity(intent);
                            }
                        } else {
                            startActivity(intent);
                        }

                        return true;
                    }
                });
    }


    /**
     * A preference value change listener that updates the preference's summary
     * to reflect its new value.
     */
    private final Preference.OnPreferenceChangeListener sBindPreferenceSummaryToValueListener = new
            Preference.OnPreferenceChangeListener() {
                @Override
                public boolean onPreferenceChange(Preference preference, Object value) {
                    String stringValue = value.toString();

                    if (preference instanceof ListPreference) {
                        // For list preferences, look up the correct display value in
                        // the preference's 'entries' list.
                        ListPreference listPreference = (ListPreference) preference;
                        int index = listPreference.findIndexOfValue(stringValue);

                        // Set the summary to reflect the new value.
                        preference.setSummary(
                                index >= 0
                                        ? listPreference.getEntries()[index]
                                        : null);

                        if (listPreference.getKey().equals(PREFS_LANGUAGE)) {
                            if (!stringValue.equals(currentLanguage)) {
                                LocaleHelper.persist(getActivity(), stringValue);
                                restartApp();
                            }
                        }

                        if (listPreference.getKey().equals(PREFS_THEME)) {
                            int theme = Integer.parseInt(stringValue);
                            preferences.edit().putInt(CURRENT_THEME, theme).apply();
                            Logger.log("TAG", "Current theme=" + SettingsPreferenceFragment.this
                                    .theme + " new theme=" + theme);
                            if (SettingsPreferenceFragment.this.theme != theme) {
                                restartApp();
                            }

//                            ((SettingsActivity) getActivity()).setApplicationTheme(theme);
                        }

                    } else {
                        // For all other preferences, set the summary to the value's
                        // simple string representation.
                        preference.setSummary(stringValue);
                    }
                    return true;
                }
            };


    /**
     * Binds a preference's summary to its value. More specifically, when the
     * preference's value is changed, its summary (line of text below the
     * preference title) is updated to reflect the value. The summary is also
     * immediately updated upon calling this method. The exact display format is
     * dependent on the type of preference.
     *
     * @see #sBindPreferenceSummaryToValueListener
     */
    private void bindPreferenceSummaryToValue(Preference preference) {
        // Set the listener to watch for value changes.
        preference.setOnPreferenceChangeListener(sBindPreferenceSummaryToValueListener);

        // Trigger the listener immediately with the preference's
        // current value.
        sBindPreferenceSummaryToValueListener.onPreferenceChange(preference,
                PreferenceManager
                        .getDefaultSharedPreferences(preference.getContext())
                        .getString(preference.getKey(), ""));
    }


    private void restartApp() {
        AppCompatActivity activity = (AppCompatActivity) getActivity();
        if (activity == null) {
            return;
        }
        final int enter_anim = android.R.anim.fade_in;
        final int exit_anim = android.R.anim.fade_out;
        activity.overridePendingTransition(enter_anim, exit_anim);
        activity.finish();
        activity.overridePendingTransition(enter_anim, exit_anim);
        activity.startActivity(activity.getIntent());
    }


}

File: app/src/main/java/com/siju/acexplorer/settings/SettingsPreferenceFragment.kt
/*
 * Copyright (C) 2017 Ace Explorer owned by Siju Sakaria
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.siju.acexplorer.settings

import android.content.Context
import android.content.SharedPreferences
import android.net.ConnectivityManager
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.preference.*
import com.google.android.play.core.appupdate.AppUpdateManager
import com.siju.acexplorer.AceApplication
import com.siju.acexplorer.R
import com.siju.acexplorer.analytics.Analytics
import com.siju.acexplorer.home.model.FavoriteHelper
import com.siju.acexplorer.logging.Logger
import com.siju.acexplorer.main.MainCommunicator
import com.siju.acexplorer.main.helper.UpdateChecker
import com.siju.acexplorer.main.model.FileConstants
import com.siju.acexplorer.main.model.root.RootUtils
import com.siju.acexplorer.main.viewmodel.MainViewModel
import com.siju.acexplorer.premium.Premium
import com.siju.acexplorer.theme.CURRENT_THEME
import com.siju.acexplorer.theme.PREFS_THEME
import com.siju.acexplorer.theme.Theme
import com.siju.acexplorer.utils.LocaleHelper
import com.siju.acexplorer.utils.NetworkHelper

const val PREFS_UPDATE = "prefsUpdate"
const val PREFS_LANGUAGE = "prefLanguage"
private const val PREFS_FULL_VERSION = "prefsUnlockFull"

class SettingsPreferenceFragment : PreferenceFragmentCompat() {

    private lateinit var mainViewModel: MainViewModel
    private lateinit var mainCommunicator: MainCommunicator
    private var preferences: SharedPreferences? = null
    private var updatePreference: Preference? = null
    private var currentLanguage: String? = null
    private var updateChecker: UpdateChecker? = null
    private var theme = 0
    private var connectivityManager : ConnectivityManager? = null
    private var networkHelper : NetworkHelper? = null
    private val handler = Handler(Looper.getMainLooper())

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is MainCommunicator) {
            updateChecker = context.getUpdateChecker()
            mainCommunicator = context
        }
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.pref_settings, rootKey)
        setupUpdatePref()
        setupUnlockFullVersionPref()
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        preferences = PreferenceManager.getDefaultSharedPreferences(context)

        setupViewModels()
        setupRootPref()
        setupLanguagePreference()
        setupThemePref()
        setupAnalyticsPref()
        setupResetFavPref()
    }

    private fun setupViewModels() {
        val activity = requireNotNull(activity)
        mainViewModel = ViewModelProvider(activity).get(MainViewModel::class.java)
        mainViewModel.premiumLiveData.observe(this, Observer {
            it?.apply {
                if (it.entitled) {
                    findPreference<Preference>(PREFS_FULL_VERSION)?.isVisible = false
                }
            }
        })
    }

    private fun setupUnlockFullVersionPref() {
        val preference = findPreference<Preference>(PREFS_FULL_VERSION)
        preference?.isVisible = !mainCommunicator.isPremiumVersion()
        preference?.setOnPreferenceClickListener {
            val activity = activity as AppCompatActivity?
            activity?.let {
                val premium = Premium(it, mainViewModel)
                premium.showPremiumDialog(it)
            }
            true
        }
    }

    private fun setupRootPref() {
        val rootPreference = findPreference(PREF_ROOT) as CheckBoxPreference?
        rootPreference?.setOnPreferenceClickListener { _ ->
            onRootPrefClicked(rootPreference.isChecked, rootPreference)
            true
        }
    }

    private fun onRootPrefClicked(newValue: Boolean, rootPreference: CheckBoxPreference) {
        if (newValue) {
            val rooted = RootUtils.hasRootAccess()
            Log.d("Settings", " rooted:$rooted")
            rootPreference.isChecked = rooted
        } else {
            rootPreference.isChecked = false
        }
    }

    private fun setupLanguagePreference() {
        val languagePreference = findPreference(PREFS_LANGUAGE) as ListPreference?
        val value = LocaleHelper.getLanguage(activity)
        languagePreference?.value = value
        currentLanguage = value
        bindPreferenceSummaryToValue(languagePreference)
    }

    private fun setupThemePref() {
        val themePreference = findPreference<ListPreference>(PREFS_THEME)
        theme = Theme.getUserThemeValue(activity!!)
        bindPreferenceSummaryToValue(themePreference)
    }

    private fun setupAnalyticsPref() {
        val analyticsPreference = findPreference(PREFS_ANALYTICS) as CheckBoxPreference?
        analyticsPreference?.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, newValue ->
            Analytics.getLogger().sendAnalytics(newValue as Boolean)
            true
        }
    }

    private fun setupResetFavPref() {
        val resetPreference = findPreference<Preference>(FileConstants.PREFS_RESET)
        resetPreference?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            resetFavorites()
            false
        }
    }

    private fun resetFavorites() {
        FavoriteHelper.resetFavorites(context)
        Toast.makeText(AceApplication.appContext, getString(R.string.msg_fav_reset), Toast
                .LENGTH_LONG).show()
    }

    private fun setupUpdatePref() {
        updatePreference = findPreference<Preference?>(PREFS_UPDATE)
        val updateAvailable = updateChecker?.isUpdateAvailable()
        if (updateAvailable == false) {
            return
        }
        networkHelper = NetworkHelper(networkChangeCallback)
        setupUpdatePrefText(NetworkHelper.isConnectedToInternet(context))
        connectivityManager = networkHelper?.getConnectivityManager(context)
        networkHelper?.registerNetworkRequest(connectivityManager)
        updatePreference?.isVisible = true
        initUpdatePrefListener()
    }

    private fun setupUpdatePrefText(networkAvailable : Boolean = true) {
        handler.post {
            when {
                updateChecker?.isUpdateDownloaded() == true -> {
                    onUpdateDownloaded()
                }
                networkAvailable && updateChecker?.isUpdateDownloading() == true -> {
                    onUpdateDownloading()
                }
                else -> {
                    updatePreference?.title = getString(R.string.update_available)
                    handleInternetSetting(networkAvailable)
                }
            }
        }
    }

    private fun initUpdatePrefListener() {
        updatePreference?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            updateClicked()
            true
        }
    }

    private fun updateClicked() {
        updateChecker?.setUpdateCallback(updateCallback)
        updateChecker?.startUpdate()
    }

    private val networkChangeCallback = object : NetworkHelper.NetworkChangeCallback {
        override fun onNetworkAvailable() {
             if (updateChecker?.isUpdateAvailable() == true) {
                 setupUpdatePrefText(true)
             }
        }

        override fun onNetworkUnavailable() {
            if (updateChecker?.isUpdateAvailable() == true) {
                setupUpdatePrefText(false)
            }
        }
    }

    private fun handleInternetSetting(networkAvailable: Boolean = true) {
        if (networkAvailable) {
            updatePreference?.summary = ""
        }
        else {
            updatePreference?.summary = getString(R.string.connect_internet_download_update)
        }
    }

    private val updateCallback = object : UpdateChecker.UpdateCallback {
        override fun onUpdateDownloaded(appUpdateManager: AppUpdateManager) {
            onUpdateDownloaded()
        }

        override fun onUpdateInstalled() {
            updatePreference?.isVisible = false
        }

        override fun onUpdateDownloading() {
            this@SettingsPreferenceFragment.onUpdateDownloading()
        }

        override fun onUpdateCancelledByUser() {
            updatePreference?.title = getString(R.string.update_available)
            handleInternetSetting(false)
        }

        override fun onUpdateSnackbarDismissed() {
        }
    }

    private fun onUpdateDownloaded() {
        updatePreference?.title = getString(R.string.update_downloaded)
        updatePreference?.summary = getString(R.string.install_update)
    }

    private fun onUpdateDownloading() {
        updatePreference?.title = getString(R.string.update_downloading)
        updatePreference?.summary = ""
    }

    /**
     * Binds a preference's summary to its value. More specifically, when the
     * preference's value is changed, its summary (line of text below the
     * preference title) is updated to reflect the value. The summary is also
     * immediately updated upon calling this method. The exact display format is
     * dependent on the type of preference.
     *
     * @see [bindPreferenceSummaryToValueListener]
     */
    private fun bindPreferenceSummaryToValue(preference: Preference?) {
        // Set the listener to watch for value changes.
        preference?.onPreferenceChangeListener = bindPreferenceSummaryToValueListener

        // Trigger the listener immediately with the preference's
        // current value.
        bindPreferenceSummaryToValueListener.onPreferenceChange(preference,
                PreferenceManager
                        .getDefaultSharedPreferences(
                                preference?.context)
                        .getString(preference?.key,
                                ""))
    }

    private val bindPreferenceSummaryToValueListener = Preference.OnPreferenceChangeListener { preference, value ->
        val stringValue = value.toString()

        if (preference is ListPreference) {
            val index = preference.findIndexOfValue(stringValue)

            // Set the summary to reflect the new value.
            preference.setSummary(
                    if (index >= 0)
                        preference.entries[index]
                    else
                        null)

            if (preference.key == PREFS_LANGUAGE) {
                onLanguagePrefChanged(stringValue)
            } else if (preference.key == PREFS_THEME) {
                onThemeChanged(stringValue)
            }

        } else {
            preference.summary = stringValue
        }
        true
    }

    private fun onThemeChanged(stringValue: String) {
        val theme = Integer.parseInt(stringValue)
        preferences?.edit()?.putInt(CURRENT_THEME, theme)?.apply()
        Logger.log("TAG", "Current theme=" + this@SettingsPreferenceFragment
                .theme + " new theme=" + theme)
        if (this@SettingsPreferenceFragment.theme != theme) {
            restartApp()
        }
    }

    private fun onLanguagePrefChanged(stringValue: String) {
        if (stringValue != currentLanguage) {
            LocaleHelper.persist(activity, stringValue)
            restartApp()
        }
    }

    private fun restartApp() {
        val activity = activity as AppCompatActivity? ?: return
        val enterAnim = android.R.anim.fade_in
        val exitAnim = android.R.anim.fade_out
        activity.overridePendingTransition(enterAnim, exitAnim)
        activity.finish()
        activity.overridePendingTransition(enterAnim, exitAnim)
        activity.startActivity(activity.intent)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        networkHelper?.unregisterNetworkRequest(connectivityManager)
    }

    companion object {
        const val PREFS_ANALYTICS = "prefsAnalytics"
        const val PREF_ROOT = "prefRooted"

    }
}